### PR TITLE
Module 3 - Added simple form on home page, on a new page. Added routes and styling.

### PR DESCRIPTION
**Pull request by sinpai**
* Added form for adding money
* Added partial for ten last records for user
* Added Controller for transactions and routes.

Deployed on Heroku - https://datarockets-finance.herokuapp.com/

### CODE REVIEW

File: app/controllers/home_page_controller.rb [hunk 1]
```diff
@@ -2,5 +2,7 @@ class HomePageController < ApplicationController
   before_action :authenticate_user!
 
   def show
+    @transaction = Transaction.new
+    @recent_records = Transaction.most_recent(current_user)
```
Comment (AleksSenkou) [inline app/controllers/home_page_controller.rb]: @sinpai what do you think about getting transactions from a user?
```ruby
user.rb

scope :most_recent_transactions -> { transactions.most_recent }
```
```ruby
transaction.rb

scope :most_recent, -> { order(created_at: :desc).limit(most_recent_count) }

private

def most_recent_count
  10
end
```

Comment (AleksSenkou) [reply app/controllers/home_page_controller.rb]: @sinpai so, after it you can use
```ruby
@recent_records = current_user.most_recent_transictions
```

File: app/views/transactions/_recent_transactions.html.slim [hunk 2]
```diff
@@ -0,0 +1,15 @@
+.col-md-8.offset-md-2
+  table.table.table-striped.table-hover
+    thead
+      tr
+        th[scope="col"] Id
+        th Sum
+        th Date
+        th Comment
```
Comment (AleksSenkou) [inline app/views/transactions/_recent_transactions.html.slim]: `Id`, `Sum`, `Date`, `Comment` to locales. Don't forget about it 😉

File: config/routes.rb [hunk 3]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
   devise_for :users, controllers: {omniauth_callbacks: 'users/omniauth_callbacks'}
   root to: 'home_page#show'
+  resources :transactions
```
Comment (AleksSenkou) [inline config/routes.rb]: @sinpai do you need all resources or just few of them? [check notes](http://guides.rubyonrails.org/routing.html#restricting-the-routes-created)

File: app/controllers/transactions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+      render 'new', notice: t('.failure')
+    end
+  end
+
+  private
+
+  def transaction_params
+    params.require(:transaction).permit(:sum, :date, :comment)
+  end
+end
```
Comment (AleksSenkou) [inline app/controllers/transactions_controller.rb:22]: @sinpai I didn't find test for this controller in `spec/controllers`.
Let's write request tests for both `new` and `create` methods

**Review by AleksSenkou — COMMENTED** [review]
I like it 👍 . You have to work on tests and locales

By the way, I've asked to add screenshots/gifs to PR's description in order to show changes

File: app/views/transactions/_recent_transactions.html.slim (hunk 2 above)
Comment (sinpai) [reply app/views/transactions/_recent_transactions.html.slim]: 👍 I was hurrying to get a review today, sorry.

File: config/routes.rb (hunk 3 above)
Comment (sinpai) [reply config/routes.rb]: Yes, I will shorten variants with only: []. 👍

File: app/controllers/transactions_controller.rb (hunk 4 above)
Comment (sinpai) [reply app/controllers/transactions_controller.rb:22]: 👍 Will implement this. Thanks

File: app/controllers/home_page_controller.rb (hunk 1 above)
Comment (sinpai) [reply app/controllers/home_page_controller.rb]: 👍

**Comment by sinpai** [comment]
http://take.ms/DDsYK - home page with simple form

**Comment by AleksSenkou** [comment]
@sinpai by the way: you can post images directly to PR's description

File: app/views/transactions/_recent_transactions.html.slim (hunk 2 above)
Comment (AleksSenkou) [reply app/views/transactions/_recent_transactions.html.slim]: @sinpai no need in a hurry 😄 
we are trying to write clean code, without startup mode 😉

File: app/views/transactions/_recent_transactions.html.slim [hunk 5]
```diff
@@ -0,0 +1,15 @@
+.col-md-8.offset-md-2
+  table.table.table-striped.table-hover
+    thead
+      tr
+        th[scope="col"]= t('.id')
+        th= ('.sum')
+        th= ('.date')
+        th= ('.comment')
```
Comment (AleksSenkou) [inline app/views/transactions/_recent_transactions.html.slim]: @sinpai is it work? I mean without calling `t` function? 😄 
```ruby
th= ('.sum')
```
```ruby
th= t('.sum')
```

File: config/routes.rb [hunk 6]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
   devise_for :users, controllers: {omniauth_callbacks: 'users/omniauth_callbacks'}
   root to: 'home_page#show'
+  resources :transactions, only: %i[new create]
```
Comment (AleksSenkou) [inline config/routes.rb:4]: 👍

File: spec/controllers/transactions_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  describe 'POST create new transaction' do
+    login_user
+    let(:transaction_params) { FactoryBot.attributes_for :transaction }
+
+    it 'has a 200 status code after create' do
+      post :create, params: {transaction: transaction_params}
+      assert_redirected_to root_path
+    end
+  end
+end
```
Comment (AleksSenkou) [inline spec/controllers/transactions_controller_spec.rb]: @sinpai 
* we have two actions in the controller: `new` and `create`. You've tested only `create`
* `create` action can 1) render `new` 2) redirect to. So we have to test both cases: creation passed and with errors

Comment (AleksSenkou) [reply spec/controllers/transactions_controller_spec.rb]: There is a cool thing for testing stuff that goes before:
```ruby
before do
  ...
end
```
such things as `login_user` should come in such blocks

check [best practisies](http://www.betterspecs.org/)

File: spec/models/transaction_spec.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
+require 'rails_helper'
+
+RSpec.describe Transaction, type: :model do
+  before(:all) do
+    12.times { FactoryBot.create(:transaction) }
```
Comment (AleksSenkou) [inline spec/models/transaction_spec.rb]: `FactoryBot.create_list(:transaction, 12)`

**Review by AleksSenkou — COMMENTED** [review]
Suggestions, good steps with tests 💪

File: app/views/transactions/_recent_transactions.html.slim (hunk 5 above)
Comment (sinpai) [reply app/views/transactions/_recent_transactions.html.slim]: It wasn't working, i've already fixed it in next PR :) Will fix it in this PR

File: spec/controllers/transactions_controller_spec.rb (hunk 7 above)
Comment (sinpai) [reply spec/controllers/transactions_controller_spec.rb]: My implementation of `login_user` function is 
```ruby
def login_user
    before(:each) do
      @request.env['devise.mapping'] = Devise.mappings[:user]
      user = FactoryBot.create(:user)
      sign_in user
    end
  end
```
So i need to just call it without `before do; end` block. Is it ok?

Comment (AleksSenkou) [reply spec/controllers/transactions_controller_spec.rb]: @sinpai yeah, it's ok 👍

File: spec/controllers/transactions_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
+require 'rails_helper'
+
+RSpec.describe TransactionsController, type: :controller do
+  let(:transaction_params) { FactoryBot.attributes_for :transaction }
+  describe 'POST create new transaction as logged user' do
```
Comment (AleksSenkou) [inline spec/controllers/transactions_controller_spec.rb]: @sinpai there should be an empty line between `let`s and `describe`

File: spec/controllers/transactions_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    it 'does not initialize a transaction not logged in' do
+      get :new
+      expect(response).to redirect_to new_user_session_path
+    end
+
+    it 'has a 302 status code after create when unlogged' do
+      post :create, params: {transaction: transaction_params}
+      expect(response.status).to eq(302)
+      assert_redirected_to new_user_session_path
+    end
```
Comment (AleksSenkou) [inline spec/controllers/transactions_controller_spec.rb]: @sinpai check best practicies once again 😉 
There is a tip for dealing with `when` statements using `context`

image

So:
```ruby
context 'when logged out' do
  context 'and after transaction creation' do
    before do
      post :create, params: {transaction: transaction_params}
    end

    it 'has a 302 status code' do
      expect(response.status).to eq(302)
    end

    it 'redirects to a new user session' do
      assert_redirected_to new_user_session_path
    end
  end
end
```

File: spec/models/transaction_spec.rb [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+  before(:all) do
+    FactoryBot.create_list(:transaction, 12)
+  end
+
+  describe 'named scope most_recent' do
+    it 'should return only 10 records' do
+      result = Transaction.most_recent
+      result.length.should eq(10)
+    end
+  end
```
Comment (AleksSenkou) [inline spec/models/transaction_spec.rb]: @sinpai a tip about methods description
image

So:
```ruby
describe '.most_recent' do
  it 'returns only 10 records' do
    expect(Transaction.most_recent.length).to eq 10
  end
end
```

**Review by AleksSenkou — APPROVED** [review]
LGTM with suggestions 💯

File: spec/models/transaction_spec.rb (hunk 11 above)
Comment (sinpai) [reply spec/models/transaction_spec.rb]: 👍

File: spec/controllers/transactions_controller_spec.rb (hunk 10 above)
Comment (sinpai) [reply spec/controllers/transactions_controller_spec.rb]: 👍

**Pull request merged**